src/include/migraphx/iota_iterator.hpp

```diff
@@ -26,134 +26,73 @@
 
 #include <migraphx/config.hpp>
 #include <migraphx/functional.hpp>
+#include <migraphx/iterator.hpp>
 #include <iterator>
-#include <type_traits>
 
 namespace migraphx {
 inline namespace MIGRAPHX_INLINE_NS {
 
 template <class F, class Iterator = std::ptrdiff_t>
-struct basic_iota_iterator
+struct basic_iota_iterator : iterator_operators<basic_iota_iterator<F, Iterator>>,
+                             iterator_types<decltype(std::declval<F>()(std::declval<Iterator>())),
+                                            std::random_access_iterator_tag>
 {
     Iterator index;
     F f;
 
-    using difference_type   = std::ptrdiff_t;
-    using reference         = decltype(f(std::declval<Iterator>()));
-    using value_type        = typename std::remove_reference<reference>::type;
-    using pointer           = typename std::add_pointer<value_type>::type;
-    using iterator_category = std::random_access_iterator_tag;
+    using reference = decltype(std::declval<F>()(std::declval<Iterator>()));
 
-    basic_iota_iterator& operator+=(int n)
+    // cppcheck-suppress uninitMemberVar
+    constexpr basic_iota_iterator() = default;
```

Comment:
This looks odd since we're silencing this cppcheck call on the default() constructor? Is this really needed then?

Reply:
Extremely weird.